pwlf/autopwlf.py

```diff
@@ -0,0 +1,232 @@
+        return input_
+
+    
+    @staticmethod
+    def _unpack_expr(func):
+        r"""
+        Extract piecewise model's parameters.
+        Parameters
+        ----------
+        func : callable object
+            Model function
+        
+        Returns
+        -------
+        k : int
+            Slope
+        b : int
+            Intercept
+        """
+        text = func.__doc__
+        l = re.findall(r'return .*\n', text)[0]
+        l = re.split(r' |\n', l)[1:-1]
+        if len(l) == 3:
+            if '*x' in l[0]:
+                k, b = eval(l[0].split('*')[0]), eval(l[1]+l[2])
+            elif '*' in l[2]:
+                k, b = eval(l[1]+l[2].split('*')[0]), eval(l[0])
+        elif len(l) == 4:
+            if '*x' in l[1]:
+                k, b = eval(l[0]+l[1].split('*')[0]), eval(l[2]+l[3])
+            elif '*x' in l[3]:
+                k, b = eval(l[2]+l[3].split('*')[0]), eval(l[0]+l[1])
+        return k, b
+
+    def fit_curve_and_guess(self, degree=16, lb=10, ub=10) -> int:
+        r"""
+        Guess lines number by using curve fit then extract roots.
+        Then initize bounds with roots-lb and roots+ub, which can
+        speed up convergence.
+        Parameters
+        ----------
+        degree : int
+            Polynomial degree to fit curve 
+        lb : int
+            Lower bound
+        ub : int
+            Up bound
+        Returns
+        -------
+        guess : int
+            The guess of piecewise lines number.
+        """
+
+        coeff = np.polyfit(self.x_data, self.y_data, degree)
+        f = np.poly1d(coeff)
+        f_d = np.polyder(f, 1)
+        xroot = np.roots(f_d)
+        xroot = list(filter(lambda x : x.imag == 0, xroot))
+        xroot = list(filter(lambda x : x >= self.x_data.min() and x <= self.x_data.max(), xroot))
+        self.n = len(xroot)+1
+        self.bounds = np.zeros([self.n-1, 2])
+        for i, e in enumerate(xroot):
+            self.bounds[i][0] = e.real-lb
+            self.bounds[i][1] = e.real+ub
+        return self.n
+
+    def piecewise_fit(self, lin_num: int = None, use_bound=True) -> None:
+        r"""
+        Use pwlf.fit() to fit model.
+        Parameters
+        ----------
+        lin_num : None, or int
+            Default lin_num=None, automatically infer segment number. 
+        use_bound : bool, optional
+            Defalut True, use bounds to speed up convergence
+        """
+        my_pwlf = pwlf.PiecewiseLinFit(self.x_data, self.y_data)
+        if use_bound == True:
+            if lin_num is None:
+                my_pwlf.fit(self.n, bounds=self.bounds, workers=-1, updating='deferred')
+            else:
+                my_pwlf.fit(lin_num, bounds=self.bounds, workers=-1, updating='deferred')
+        else:
+            if lin_num is None:
+                my_pwlf.fit(self.n, workers=-1, updating='deferred')
+            else:
+                my_pwlf.fit(lin_num, workers=-1, updating='deferred')
```

Comment:
Instead of `my_pwlf.fit()` try `fit_guess(x_root, bounds=self.bounds)`. I suspect this will be good enough, and much faster than the even the parallel differential evolution!